bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/ComputeAndValidateHashFromBigtableDoFn.java

```diff
@@ -0,0 +1,232 @@
+  private final ValueProvider<String> tableName;
+  private final ValueProvider<String> projectId;
+  private final ValueProvider<String> sourceHashDir;
+
+  private final TableHashWrapperFactory tableHashWrapperFactory;
+
+  // Counter for reporting matching and mismatching ranges. Names are similar to HBase sync-table
+  // job.
+  private final Counter matches = Metrics.counter("cbt-dataflow-validate", "ranges_matched");
+  private final Counter mismatches = Metrics.counter("cbt-dataflow-validate", "ranges_not_matched");
+
+  public ComputeAndValidateHashFromBigtableDoFn(SyncTableOptions options) {
+    super(TemplateUtils.BuildSyncTableConfig(options));
+    this.tableName = options.getBigtableTableId();
+    // Create a local copy of ValueProviders, PipelineOptions are not serializable.
+    projectId = options.getBigtableProject();
+    sourceHashDir = options.getHashTableOutputDir();
+    tableHashWrapperFactory = new TableHashWrapperFactory();
+  }
+
+  @VisibleForTesting
+  ComputeAndValidateHashFromBigtableDoFn(
+      CloudBigtableConfiguration config,
+      ValueProvider<String> tableName,
+      ValueProvider<String> projectId,
+      ValueProvider<String> sourceHashDir,
+      TableHashWrapperFactory factory) {
+    super(config);
+    this.tableName = tableName;
+    this.tableHashWrapperFactory = factory;
+    this.sourceHashDir = projectId;
+    this.projectId = sourceHashDir;
+  }
+
+  @ProcessElement
+  public void processElement(ProcessContext context) throws Exception {
+    // BufferedHadoopHashTableSource generates only 1 item per groupby key, but iterate just in
+    // case.
+    for (List<RangeHash> rangeHashes : context.element().getValue()) {
+      if (rangeHashes.isEmpty()) {
+        // No rows ranges found, return;
+        return;
+      }
+
+      ImmutableBytesWritable rangeStartInclusive = rangeHashes.get(0).startInclusive;
+      ImmutableBytesWritable rangeEndExclusive =
+          rangeHashes.get(rangeHashes.size() - 1).stopExclusive;
+
+      BigtableResultHasher resultHasher = new BigtableResultHasher();
+      resultHasher.startBatch(rangeStartInclusive);
+
+      // Since all the row-ranges are sorted in HashTable's data files, 1 big scan can be used
+      // to read all the row ranges. Parallelism is achieved by splitting the HashTable's data
+      // files into smaller bundle of row-ranges in GroupBy.
+      ResultScanner scanner =
+          createBigtableScan(rangeStartInclusive.copyBytes(), rangeEndExclusive.copyBytes());
+
+      Iterator<RangeHash> rangeHashIterator = rangeHashes.iterator();
+      long numRows = 0;
+
+      RangeHash currentRangeHash = rangeHashIterator.next();
+
+      // Process each row and validate hashes
+      for (Result result : scanner) {
+        numRows++;
+        if (numRows % 10_000 == 0) {
+          // Heartbeat in logs in case a large scan gets hung.
+          DOFN_LOG.debug("Processed " + numRows + " rows ");
```

Comment:
would a counter be a better choice?

Reply:
I am not sure if counter will be better. We already have a Success/fail counter that can be monitored. This is for debugging if one worker had an issue, logs were filtered on a worker and you see these logs.